### PR TITLE
Add streaming support for Coindirect

### PR DESCRIPTION
**Pull request by donald-jackson**
Please see pull request for added streaming Order books and trades for Coindirect.

This depends on the current latest XChange with Coindirect integration (currently 4.3.9-SNAPSHOT) - please review/edit if you wish to the appropriate release versions.

### CODE REVIEW

File: xchange-coindirect/pom.xml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.knowm.xchange</groupId>
+            <artifactId>xchange-coindirect</artifactId>
+            <version>${xchange.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>info.bitrich.xchange-stream</groupId>
+            <artifactId>service-netty</artifactId>
+            <version>4.3.3-SNAPSHOT</version>
```
Comment (declan94) [inline xchange-coindirect/pom.xml]: It's better to use maven variable instead of the hard-coded version.

**Comment by declan94** [comment]
Is there anyone else who uses Coindirect can help test this PR?

File: xchange-coindirect/src/main/java/info/bitrich/xchangestream/coindirect/service/CoindirectStreamingMarketDataService.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    @Override
+    public Observable<Trade> getTrades(CurrencyPair currencyPair, Object... args) {
+        String channelName = "ticker-" + CoindirectAdapters.toSymbol(currencyPair);
+        return service.subscribeChannel(channelName).map(s -> {
+            CoindirectTradeEvent coindirectTradeEvent = objectMapper.readValue(s, CoindirectTradeEvent.class);
+            return new Trade(Order.OrderType.BID, coindirectTradeEvent.amount, currencyPair, coindirectTradeEvent.price, new Date(coindirectTradeEvent.timestamp), null);
+        });
+    }
+
+    public Observable<Trade> getAccountTrades() {
```
Comment (badgerwithagun) [inline xchange-coindirect/src/main/java/info/bitrich/xchangestream/coindirect/service/CoindirectStreamingMarketDataService.java]: Rename to `getTrades` for consistency

**Comment by badgerwithagun** [comment]
Executive decision: when reviewing PRs for new exchanges, or exchanges where there are no visible other users, I am now reviewing them in the same way that the XChange project does: review the code for obvious issues or inconsistencies with the rest of the library, and otherwise assume it works for the author or they wouldn't have submitter it, so it will be useful to someone.

On that basis, this looks basically fine. @donald-jackson , if you can just follow up to the two comments above, I'll merge this.

File: xchange-coindirect/pom.xml (hunk 1 above)
Comment (donald-jackson) [reply xchange-coindirect/pom.xml]: Oops, apologies

File: xchange-coindirect/src/main/java/info/bitrich/xchangestream/coindirect/service/CoindirectStreamingMarketDataService.java (hunk 2 above)
Comment (donald-jackson) [reply xchange-coindirect/src/main/java/info/bitrich/xchangestream/coindirect/service/CoindirectStreamingMarketDataService.java]: Thanks

**Comment by donald-jackson** [comment]
@badgerwithagun yes thank you, I am using this in production at this stage

**Comment by badgerwithagun** [comment]
Hi @donald-jackson - I'm just catching up on some old PRs and would like to get this merged for you.  Apologies that it's taken so long. If I give it a final review sweep, will you be able to apply the fixes?

**Comment by badgerwithagun** [comment]
I'm going to close this PR unmerged in a few days as there seems to be no interest in getting it over the line.

**Comment by donald-jackson** [comment]
Apologies for the delay I've been away over the holidays. Yes I'll apply fixes if you give it a final sweep, thanks.

File: service-centrifugo/src/main/java/info/bitrich/xchangestream/service/centrifugo/CentrifugoStreamingService.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+    public Completable disconnect() {
+        return Completable.create(completable -> {
+            LOG.info("Disconnecting Centrifugo");
+            centrifugo.disconnect();
+            completable.onComplete();
+        });
+    }
+
+    public Observable<String> subscribeChannel(String channelName) {
```
Comment (badgerwithagun) [inline service-centrifugo/src/main/java/info/bitrich/xchangestream/service/centrifugo/CentrifugoStreamingService.java]: This code isn't thread-safe. Are you confident it won't be called by more than one thread concurrently?

File: service-centrifugo/src/main/java/info/bitrich/xchangestream/service/centrifugo/CentrifugoStreamingService.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        LOG.info("Subscribing to channel {}.", channelName);
+        Observable<String> observable = Observable.<String>create(e -> {
+            if (state != State.CONNECTED) {
+                LOG.error("Centrifugo not connected.");
+                e.onError(new NotConnectedException());
+                return;
+            }
+
+            LOG.debug("Sending subscribe for {}", channelName);
+            centrifugo.subscribe(new SubscriptionRequest(channelName));
```
Comment (badgerwithagun) [inline service-centrifugo/src/main/java/info/bitrich/xchangestream/service/centrifugo/CentrifugoStreamingService.java]: Will there be any issues if you end up subscribing to the same channel multiple times?

File: service-centrifugo/src/main/java/info/bitrich/xchangestream/service/centrifugo/CentrifugoToken.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
+package info.bitrich.xchangestream.service.centrifugo;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+
+public class CentrifugoToken {
```
Comment (badgerwithagun) [inline service-centrifugo/src/main/java/info/bitrich/xchangestream/service/centrifugo/CentrifugoToken.java]: Does this need to be public visibility?

File: xchange-coindirect/pom.xml (hunk 1 above)
Comment (badgerwithagun) [reply xchange-coindirect/pom.xml]: Should use the parent project version

File: xchange-coindirect/src/main/java/info/bitrich/xchangestream/coindirect/CoindirectStreamingExchange.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+    private CentrifugoToken getUserToken() {
+        CoindirectTokenService coindirectTokenService = new CoindirectTokenService("https://user-token.coindirect.com");
+        try {
+            coindirectUserToken = coindirectTokenService.getUserToken();
+            CentrifugoToken centrifugoToken = new CentrifugoToken(coindirectUserToken.user, coindirectUserToken.timestamp, coindirectUserToken.token);
+            centrifugoToken.setUrl(coindirectUserToken.url);
+            return centrifugoToken;
+        } catch(IOException ignored) {
+            return null;
```
Comment (badgerwithagun) [inline xchange-coindirect/src/main/java/info/bitrich/xchangestream/coindirect/CoindirectStreamingExchange.java]: Nulls aren't safely handled by `CentrifugoStreamingService`.

File: xchange-coindirect/src/main/java/info/bitrich/xchangestream/coindirect/CoindirectStreamingService.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+import info.bitrich.xchangestream.service.centrifugo.CentrifugoStreamingService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+
+public class CoindirectStreamingService extends CentrifugoStreamingService {
+    private static final Logger LOG = LoggerFactory.getLogger(CoindirectStreamingService.class);
+
+    List<ProductSubscription> subscriptionList;
```
Comment (badgerwithagun) [inline xchange-coindirect/src/main/java/info/bitrich/xchangestream/coindirect/CoindirectStreamingService.java]: Not clear why this is here.

File: xchange-coindirect/src/main/java/info/bitrich/xchangestream/coindirect/service/CoindirectStreamingMarketDataService.java [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    private CoindirectMarketDataServiceRaw coindirectMarketDataServiceRaw;
+    private CoindirectAccountServiceRaw coindirectAccountServiceRaw;
+
+    public CoindirectStreamingMarketDataService(CoindirectStreamingService service, Exchange exchange) {
+        this.service = service;
+        this.exchange = exchange;
+    }
+
+    private CoindirectMarketDataServiceRaw getCoindirectMarketDataServiceRaw() {
+        if(coindirectMarketDataServiceRaw == null) {
```
Comment (badgerwithagun) [inline xchange-coindirect/src/main/java/info/bitrich/xchangestream/coindirect/service/CoindirectStreamingMarketDataService.java]: Thread safety

File: xchange-coindirect/src/main/java/info/bitrich/xchangestream/coindirect/service/CoindirectStreamingMarketDataService.java [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+            coindirectMarketDataServiceRaw = new CoindirectMarketDataServiceRaw(exchange);
+        }
+        return coindirectMarketDataServiceRaw;
+    }
+
+    private CoindirectAccountServiceRaw getCoindirectAccountServiceRaw() {
+        if(coindirectAccountServiceRaw == null) {
```
Comment (badgerwithagun) [inline xchange-coindirect/src/main/java/info/bitrich/xchangestream/coindirect/service/CoindirectStreamingMarketDataService.java]: Thread safety

File: xchange-coindirect/src/main/java/info/bitrich/xchangestream/coindirect/service/CoindirectStreamingMarketDataService.java [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+            coindirectAccountServiceRaw = new CoindirectAccountServiceRaw(exchange);
+        }
+        return coindirectAccountServiceRaw;
+    }
+
+    @Override
+    public Observable<OrderBook> getOrderBook(CurrencyPair currencyPair, Object... args) {
```
Comment (badgerwithagun) [inline xchange-coindirect/src/main/java/info/bitrich/xchangestream/coindirect/service/CoindirectStreamingMarketDataService.java]: This code permits multiple parallel subscriptions to the same channel but is not thread safe as a result of shared state

File: xchange-coindirect/src/main/java/info/bitrich/xchangestream/coindirect/service/CoindirectStreamingMarketDataService.java (hunk 2 above)
Comment (badgerwithagun) [reply xchange-coindirect/src/main/java/info/bitrich/xchangestream/coindirect/service/CoindirectStreamingMarketDataService.java]: This belongs on a `StreamingTradeService` implementation

**Comment by badgerwithagun** [comment]
This PR is scheduled for closure due to being stale. We are shortly going to merge this project with XChange so need to close all outstanding PRs.

**Pull request closed without merging**